Sources/AsyncAlgorithms/AsyncBroadcastSequence.swift

```diff
@@ -0,0 +1,246 @@
+            self.terminal = .finished
+            continuation.resume(returning: .failure(error))
+          case .finished:
+            continuation.resume(returning: .success(nil))
+          }
+          self.continuation = nil
+        }
+      }
+      
+      mutating func cancel() {
+        buffer.removeAll()
+        terminal = .finished
+        drain()
+      }
+      
+      mutating func next(_ continuation: UnsafeContinuation<Result<Element?, Error>, Never>) {
+        assert(self.continuation == nil) // presume that the sides are NOT sendable iterators...
+        self.continuation = continuation
+        drain()
+      }
+      
+      mutating func emit(_ result: Result<Element?, Error>) {
+        switch result {
+        case .success(let element):
+          if let element {
+            buffer.append(element)
+          } else {
+            terminal = .finished
+          }
+        case .failure(let error):
+          terminal = .failure(error)
+        }
+        drain()
+      }
+    }
+    
+    var id = 0
+    var sides = [Int: Side]()
+    
+    init() { }
+    
+    mutating func establish() -> Int {
+      defer { id += 1 }
+      sides[id] = Side()
+      return id
+    }
+    
+    static func establish(_ state: ManagedCriticalState<State>) -> Int {
+      state.withCriticalRegion { $0.establish() }
+    }
+    
+    mutating func cancel(_ id: Int) {
+      if var side = sides.removeValue(forKey: id) {
+        side.cancel()
+      }
+    }
+    
+    static func cancel(_ state: ManagedCriticalState<State>, id: Int) {
+      state.withCriticalRegion { $0.cancel(id) }
+    }
+    
+    mutating func next(_ id: Int, continuation: UnsafeContinuation<Result<Element?, Error>, Never>) {
+      sides[id]?.next(continuation)
+    }
+    
+    static func next(_ state: ManagedCriticalState<State>, id: Int) async -> Result<Element?, Error> {
+      await withUnsafeContinuation { continuation in
+        state.withCriticalRegion { $0.next(id, continuation: continuation) }
+      }
+    }
+    
+    mutating func emit(_ result: Result<Element?, Error>) {
+      for id in sides.keys {
+        sides[id]?.emit(result)
+      }
+    }
+    
+    static func emit(_ state: ManagedCriticalState<State>, result: Result<Element?, Error>) {
+      state.withCriticalRegion { $0.emit(result) }
+    }
+  }
+  
+  struct Iteration {
+    enum Status {
+      case initial(Base)
+      case iterating(Task<Void, Never>)
+      case terminal
+    }
+    
+    var status: Status
+    
+    init(_ base: Base) {
+      status = .initial(base)
+    }
+    
+    static func task(_ state: ManagedCriticalState<State>, base: Base) -> Task<Void, Never> {
+      Task {
+        do {
+          for try await element in base {
+            State.emit(state, result: .success(element))
+          }
+          State.emit(state, result: .success(nil))
+        } catch {
+          State.emit(state, result: .failure(error))
+        }
+      }
+    }
```

Comment:
One thing I just noticed after the discussion in the forums around deferred. This is actually creating multiple upstream iterators if I understand this code correctly. IMO we really need to avoid this otherwise this algorithm is not capable of transforming a unicast `AsyncSequence` into a broadcasted one. @phausler please correct if I am misunderstanding the code here

Reply:
no, this is creating 1 singular task to iterate so only 1 upstream iterator is made.

Reply:
I'm clearly being dense, because it also looks to me like this would create a new Task, at least each time `broadcast` is called. And for each `Task` there would be a separate iterator of the base sequence.